### DISCUSSION TITLE
RISC-V support

### DISCUSSION THREAD

**Discussion by HochuOlivie** [upx/upx#793] [Ideas]
Hello, we are trying to make the RISC-V support for your project. Could you please share the workflow for implementing a new architecture?

**Comment by jreiser** [comment]
0. The register usage conventions and subroutine calling sequence for C-language code must be documented in a .pdf URL that has public access permissions.
1. The command-line invocation arguments to compile and link C-language and assembly-language code using both **gcc** and **clang** cross compilers running under Linux on x86_64 must be documented.
2. Make sure that `/usr/include/elf.h` has the correct constants for your architecture.
3. In `upx/src/stub/Makefile`, use the entries that contain `amd64-linux.elf-*` as a guide to construct corresponding entries (and corresponding `*.S` files) for `riscv64-linux.elf-*`.
4. Get the `upx/upx-stubtools` package and install it.  Then `cd upx/src/stub; make`
5.  Add EM_RISCV64 to `upx/src/p_lx_elf.cpp`, using all `EM_X86_64` as a guide.
7. `cd upx/build/debug; cmake --build .`, possibly with `--verbose`.

**Comment by markus-oberhumer** [comment]
The first step would be to decide on a cross-compiler that could be added to upx-stubtools. An official binary from Debian would be preferred.

**Comment by markus-oberhumer** [comment]
Looks like Debian has cross riscv binutils-2.42 and gcc-13.2:
- https://packages.debian.org/trixie/binutils-riscv64-linux-gnu
- https://packages.debian.org/trixie/gcc-13-riscv64-linux-gnu

Unfortunately the gcc package needs glibc-2.36 which is not available in GitHub Ubuntu-22.04 CI Actions.

But then, Ubuntu-24.04 should be available soon.

**Comment by yilingqinghan** [comment]
Hi, it’s been over a year since this thread started. Has RISC-V support for UPX progressed since then? I’d appreciate it if there’s any branch, patch, or update available to test or build with. 😄

**Reply by jreiser** [reply]
The attention of UPX developers has been focused on other $ARCH. Ubuntu 24.04 now is available, so that roadblock for development via Continuous Integration has been removed.  Still missing:
1. Specific downloadable URL of documentation of machine architecture and instruction set.
2. Specific downloadable URL of ABI for user<-> kernel communication via system calls.
3. Specific downloadable URL of register usage by compiled code: arguments in, result out, registers that must be saved, registers that must not be used by code in user mode.
4. Specific URLs of hardware kit, analogous to RaspberryPi in functionality, cost, availability, and administration: single-board computer, case, power supply. Yes, we know and use qemu; but there is no substitute for real hardware, because documentation and emulators have bugs.  UPX runs into (and helps fix) a qemu bug something like once per year, and it is extremely disruptive.

**Comment by jreiser** [comment]
For programming UPX de-compression, then RISCV-64 architecture has a couple unfortunate omissions.  There is no instruction ZEXT.W (zero extend 32 bits to 64 bits), and C.JAL (16-bit instruction to Jump And Link within +/- 1KB) is RISCV-32 only.  Of course it's a Turing machine, so anything is possible, but the costs might be a lot more than desired. In particular, the cost is around +25% in space for the de-compressor of around 3KB.  Nearly every subroutine call spans at most 1KB, and there are many lexical calls.

The compiled C-code portion of the de-compressor is littered with `slliw r2,r1,32; srliw r2,r2,32` because there is no `zextw r2,r1`.  The convention of 32-bit data always being sign-extended in a 64-bit register is costly.  The repeated assertion in RISC-V Instruction Set Manual that "sign extension of 32-bit unsigned causes no problems" is INCORRECT.  Quite a lot of C code uses `unsigned int` (32-bits) to deal with external data that is 32-bit unsigned, and for over 50 years programmers have relied on `unsigned` meaning "has NO sign bit": the bits in a register are a viewport into an infinite sequence of bits which are otherwise ZERO.  Large pieces of code for both graphics and cryptography rely on this, but RISC-V  hardware, ABI convention, and gcc implementation do not support it.

The current software support also is not so good. `riscv64-linux-gnu-gcc (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0` on Ubuntu 24.04 does not let me write `C.` instructions, so I cannot say that I know I want the 16-bit flavors instead of the 32-bit flavors, or even the `JAL` without a preceding `AUIPC`.  Automatic R_RISCV_RELAX relocation is too late, confusing and bug-ridden, and not enough.  I really do know more about upx code than the build system can figure out by itself.

The immediate problem is that utilities such as objstrip sometimes delete the symbol that is referenced only by a R_RISCV_RELAX relocation.  So first I have to write a utility to remove the RELAX; `/usr/bin/ld -r` has no option to do so.

**Comment by jreiser** [comment]
It's time for a market survey. Among the RiscV-64 hardware variants in today's delivered chips, what fraction support the Compressed instruction set (16-bit encoding of many instructions, in addition to the usual 32-bit encoding), and what fraction support unaligned memory access (ability to fetch and store 32-bit words at an arbitrary memory address, not just an address that is divisible by 4)? In both space and time, the upx de-compressor varies several percent depending on the hardware variant.

**Comment by markus-oberhumer** [comment]
RVA22U64 ?

**Comment by jreiser** [comment]
[RVA22U64](https://github.com/riscv/riscv-profiles/blob/main/src/profiles.adoc) requires extension Ziccism, which requires that main memory support misaligned loads and stores.  However,
```
This [Ziccism] is a new extension name for this feature. Even though mandated,
misaligned loads and stores might execute extremely slowly. Standard software
distributions should assume their existence only for correctness, not for performance. 
```
In other words, misaligned access could be implemented by a trap, which must be handled by the hardware or the operating system, transparently to any application.  So each misaligned access could be (say) 50 times slower.  UPX de-compression requires it only in one place (and 1 more if unfiltering), and the user-level code is only 10 times slower than an aligned 4-byte access.  So I'll do it byte-by-byte in software.

**Comment by jreiser** [comment]
How should UPX achieve multi-processor propagation of the de-compressed instructions?  The FENCE.I instruction makes all data stores visible to any instruction fetch, but only on the CPU core ("hart") that issued the FENCE.I.  From riscv-spec.pdf p.32 (Document Version 20190608-Base-Ratified):
```
Because FENCE.I only orders stores with a hart’s own instruction fetches, application code
should only rely upon FENCE.I if the application thread will not be migrated to a different hart.
The EEI can provide mechanisms for efficient multiprocessor instruction-stream synchronization.
```
So what mechanisms are actually used by EEI (Execution Environment Implementation: i.e., real chips and OS)?

**Comment by cjvth** [comment]
Hello. I just noticed some updates here. Last year I was working on riscv64 support for UPX (without shared libraries) as my bachelor's diploma. It seems to work, I couldn't find any cases when it fails. But I felt like I must reformat code and add tests before submitting a PR, and then couldn't find time.

Not sure if my help would be needed since I'm not that experienced. Also, my assembly code is mostly just replacing x86 instructions with riscv, and then trying to optimize the size. However, I'm open to help where I can.

Our repository is https://github.com/ylab-nsu/upx/ Also, there is another student that tries to implement riscv32 support, but he chose a very different path to do it.

**Comment by jreiser** [comment]
Alright, peope: how should UPX ensure that de-compressed code is seen by any following instruction fetches?
The default is to `write` the entire de-compressed region to `/dev/null`, because an operating system kernel must make sure that an I/O device sees all previous stores to memory.  But such a write is slow.  ARM64 can do the required flushes from user mode, using a 30-line subroutine; see `src/stub/src/arch/arm64/v8/macros.S `for the `sync_cache` macro.  Some riscv64 manual page says that there is a subroutine `__riscv_flush_icache` but I cannot find it.  (By itself that is insufficient, because the data cache must be flushed first.)

**Comment by jreiser** [comment]
riscv64 main programs are functional!
```
2025-12-13:
  Builds (Actions page): https://github.com/upx/upx/actions/runs/20197681069
  ucl/nrv compression methods work for riscv64 main programs.
    Tested under Ubuntu-24.04 on amd64 (x86_64) using:
       sudo apt install crossbuild-essential-riscv64
       qemu-riscv64 -strace -g 1234 -L /usr/riscv64-linux-gnu <program>  &
       gdb-multiarch -q <program>;   target remote :1234;   continue
  lzma stub gets SIGSEGV; suspect gcc compiler bug.
  Support for compressing shared libraries is not yet implemented.
  Filtering of AUIPC instruction (for better compression) is not yet implemented.
```

**Comment by jreiser** [comment]
2025-12-24:
riscv64 shared libraries work! Tested under qemu-riscv64 on Ubuntu 24.04 x86_64.

Actions page https://github.com/upx/upx/actions/runs/20496425112 has builds.

Pre-compression filtering (enables tighter compression of AUIPC instructions) is NotYetImplemented.  But all NRV/UCL and Lzma compression methods pass their continuous integration tests.

**Comment by jreiser** [comment]
2026-01-07:
UPX release 5.1.0 includes filtering to increase compression of AUIPC instructions.

Users please report experiences with UPX support for riscv64-linux.
